### PR TITLE
fix(core): fixed race condition releasing (and potentially leaking) table reader

### LINKED ISSUE #1745: Flaky ReaderPoolTest.testGetAndCloseRace test
### Describe the bug

Flapping test:

```
2021-12-28T09:44:59.4359640Z 2021-12-28T09:44:58.884459Z I i.q.c.AbstractCairoTest Tearing down test ReaderPoolTest#testGetAndCloseRace
2021-12-28T09:44:59.4362640Z 2021-12-28T09:44:58.902433Z E i.q.TestListener ***** Test Failed *****io.questdb.cairo.pool.ReaderPoolTest.testGetAndCloseRace : 
2021-12-28T09:44:59.4364750Z java.lang.AssertionError: [15,12,16,17,18] expected:<3> but was:<6>
2021-12-28T09:44:59.4366400Z    at org.junit.Assert.fail(Assert.java:89)
2021-12-28T09:44:59.4367700Z    at org.junit.Assert.failNotEquals(Assert.java:835)
2021-12-28T09:44:59.4369500Z    at org.junit.Assert.assertEquals(Assert.java:647)
2021-12-28T09:44:59.4374600Z    at io.questdb.test.tools.TestUtils.assertMemoryLeak(TestUtils.java:418)
2021-12-28T09:44:59.4376290Z    at io.questdb.cairo.pool.ReaderPoolTest.assertWithPool(ReaderPoolTest.java:877)
2021-12-28T09:44:59.4378090Z    at io.questdb.cairo.pool.ReaderPoolTest.assertWithPool(ReaderPoolTest.java:873)
2021-12-28T09:44:59.4380670Z    at io.questdb.cairo.pool.ReaderPoolTest.testGetAndCloseRace(ReaderPoolTest.java:382)
2021-12-28T09:44:59.4384810Z    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2021-12-28T09:44:59.4386240Z    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2021-12-28T09:44:59.4387810Z    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2021-12-28T09:44:59.4393480Z    at java.lang.reflect.Method.invoke(Method.java:566)
2021-12-28T09:44:59.4395510Z    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
```

Observed in https://dev.azure.com/questdb/questdb/_build/results?buildId=4386&view=logs&j=3fa4b4be-6db3-54fa-4432-6a83730bebb7&t=bb90c814-cfaa-5557-8498-4851227a92fe

Please do not confuse with https://github.com/questdb/questdb/issues/1513, which is fixed

### To reproduce

Run this test to failure

### Expected Behavior

_No response_

### Environment

```markdown
- **QuestDB version**:
- **OS**:
- **Browser**:
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bluestreak01**
closes #1745

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|1|1|100.00%|

File: core/src/main/java/io/questdb/cairo/pool/ReaderPool.java [hunk 1]
```diff
@@ -317,12 +316,7 @@ private boolean returnToPool(R reader) {
 
             e.releaseTimes[index] = clock.getTicks();
             Unsafe.arrayPutOrdered(e.allocations, index, UNALLOCATED);
-
-            // todo: there is a race condition between this method and
-            //   releaseAll() when the latter shuts down the pool. I thought of adding a version counter
-            //   that each method will attempt to increment thus recognising race condition and
-            //   taking action
-            return true;
+            return !isClosed();
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/cairo/pool/ReaderPool.java:319]: There is another failure that may be seen in `testGetAndCloseRace` after this change:
```
java.lang.IllegalStateException: fd 19 is already closed!
	at io.questdb/io.questdb.std.Files.auditClose(Files.java:61)
	at io.questdb/io.questdb.std.Files.close(Files.java:94)
	at io.questdb/io.questdb.std.FilesFacadeImpl.close(FilesFacadeImpl.java:44)
	at io.questdb/io.questdb.cairo.vm.MemoryCMRImpl.close(MemoryCMRImpl.java:56)
	at io.questdb/io.questdb.std.Misc.free(Misc.java:45)
	at io.questdb/io.questdb.cairo.TableReaderMetadata.close(TableReaderMetadata.java:187)
	at io.questdb/io.questdb.std.Misc.free(Misc.java:45)
	at io.questdb/io.questdb.cairo.TableReader.close(TableReader.java:163)
	at io.questdb/io.questdb.cairo.pool.ReaderPool$R.close(ReaderPool.java:370)
	at io.questdb/io.questdb.cairo.pool.ReaderPool.closeReader(ReaderPool.java:271)
	at io.questdb/io.questdb.cairo.pool.ReaderPool.releaseAll(ReaderPool.java:240)
	at io.questdb/io.questdb.cairo.pool.AbstractPool.closePool(AbstractPool.java:82)
	at io.questdb/io.questdb.cairo.pool.ReaderPool.closePool(ReaderPool.java:217)
	at io.questdb/io.questdb.cairo.pool.AbstractPool.close(AbstractPool.java:57)
	at io.questdb/io.questdb.cairo.pool.ReaderPoolTest.lambda$testGetAndCloseRace$12(ReaderPoolTest.java:390)
```

To reproduce it, you may increase the loop limit in `testGetAndCloseRace` from 100 to 1000 and then use "Run until failure" run configuration parameter in IDEA. It takes around 300-400 runs on my machine to reproduce the failure.

Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/pool/ReaderPool.java:319]: Thanks Andrey, this makes sense. I know what this is - similar behaviour as we observed on WriterPool

Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/pool/ReaderPool.java:319]: should be fixed in https://github.com/questdb/questdb/pull/1746/commits/47b198c1f4ad612ce29ba95a6e7e3948836d79bf

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 4 / 4 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 3 / 4 (75.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|3|4|75.00%|

File: core/src/main/java/io/questdb/cairo/pool/ReaderPool.java (hunk 1 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/cairo/pool/ReaderPool.java:319]: Yes, the race seems to be fixed now. Thanks.

**Review by puzpuzpuz — APPROVED** [review]

**Pull request merged**